### PR TITLE
TSK-1957: groups are included in the domain aggregation

### PR DESCRIPTION
**Pull request by norman-schmidt**
- added schema GROUP_INFO
- added group related methods to UserMapper / UserMapperSqlProvider
- User API extended with groups
- UserInfoRefreshJob includes groups

---
Release Notes:

```
**Breaking**
* TSK-1957: This TASKANA version requires a database schema update (which is backwards compatible)

** New **
* TSK-1957: user model now has 'groups' attribute, which are included in the domain aggregation when getting a user
```
 
### For the submitter:
- [x] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- https://taskana.atlassian.net/wiki/spaces/TAS/pages/2735734785/Data+Model+v5.2.0 (new UML image has to be provided)
- [ ] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- https://sonarcloud.io/summary/new_code?id=norman-schmidt_taskana&branch=TSK-1957-groups
- [ ] After integration of the PR, I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [x] Commit message format → TSK-XXX: Your commit message.
- [x] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [x] SonarCloud analysis meets our standards
- [x] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: common/taskana-common/src/main/resources/sql/db2/taskana-schema-db2.sql [hunk 1]
```diff
@@ -342,6 +342,12 @@ CREATE TABLE USER_INFO
     PRIMARY KEY (USER_ID)
 );
 
+CREATE TABLE GROUP_INFO
+(
+    USER_ID     VARCHAR(32) NOT NULL,
+    GROUP_ID    VARCHAR(256) NOT NULL,
```
Comment (mustaphazorgati) [inline common/taskana-common/src/main/resources/sql/db2/taskana-schema-db2.sql:348]: Is a length of 256 characters necessary?

File: common/taskana-common/src/main/resources/sql/h2/taskana_schema_update_5.1.0_to_5.2.0_h2.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+-- this script updates the TASKANA database schema from version 5.1.0 to version 5.2.0.
```
Comment (mustaphazorgati) [inline common/taskana-common/src/main/resources/sql/h2/taskana_schema_update_5.1.0_to_5.2.0_h2.sql:1]: did you verify the update scripts?

File: lib/taskana-core-test/src/test/java/acceptance/user/UserServiceAccTest.java [hunk 3]
```diff
@@ -236,6 +243,11 @@ void should_UpdateUserInDatabase_When_IdExisting() throws Exception {
       User userToUpdate = randomTestUser().buildAndStore(userService);
 
       userToUpdate.setFirstName("Anton");
+      userToUpdate.setGroups(
+          Set.of(
+              "cn=ksc-users,cn=groups,OU=Test,O=TASKANA",
+              "cn=Organisationseinheit KSC 2,cn=Organisationseinheit"
+                  + " KSC,cn=organisation,OU=Test,O=TASKANA"));
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/user/UserServiceAccTest.java:250]: should we make a dedicated test for the groups attribute?

File: lib/taskana-core-test/src/test/java/acceptance/user/UserServiceAccTest.java [hunk 4]
```diff
@@ -361,7 +373,26 @@ void should_ReturnOneDomain_When_UserHasSufficientMinimalPermissionsToAssignDoma
...
+          user.getId(), workbasket, WorkbasketPermission.OPEN, WorkbasketPermission.READ);
+
+      User userInDatabase = userService.getUser(user.getId());
+
+      assertThat(userInDatabase.getDomains()).containsExactly(workbasket.getDomain());
+    }
+
+    @WithAccessId(user = "user-1-1")
+    @Test
+    void should_ReturnOneDomain_When_GroupsHasSufficientMinimalPermissionsToAssignDomains()
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/user/UserServiceAccTest.java:386]:
```suggestion
    void should_ReturnOneDomain_When_GroupHasSufficientMinimalPermissionsToAssignDomains()
```

File: lib/taskana-core-test/src/test/java/acceptance/user/UserServiceAccTest.java [hunk 5]
```diff
@@ -361,7 +373,26 @@ void should_ReturnOneDomain_When_UserHasSufficientMinimalPermissionsToAssignDoma
...
+        throws Exception {
+      User user = randomTestUser().buildAndStore(userService, "businessadmin");
+      Workbasket workbasket =
+          defaultTestWorkbasket().buildAndStore(workbasketService, "businessadmin");
+      createAccessItem(
+          user.getGroups().iterator().next(),
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/user/UserServiceAccTest.java:392]: Instead of using random values I prefer to set the group explicitly

File: lib/taskana-core-test/src/test/java/acceptance/user/UserServiceAccTest.java [hunk 6]
```diff
@@ -392,6 +424,32 @@ void should_ReturnEmptyDomains_When_UserHasSufficientPermissionsWhichThenGetRevo
...
     }
 
+    @WithAccessId(user = "businessadmin")
+    @Test
+    void should_ReturnEmptyDomains_When_GroupHasSufficientPermissionsAndThenGroupIsUpdated()
+        throws Exception {
+      User user = randomTestUser().buildAndStore(userService);
+      Workbasket workbasket = defaultTestWorkbasket().buildAndStore(workbasketService);
+      createAccessItem(
+          user.getGroups().iterator().next(),
```
Comment (mustaphazorgati) [inline lib/taskana-core-test/src/test/java/acceptance/user/UserServiceAccTest.java:434]: I prefer setting the groups attribute explicitly

File: lib/taskana-core/src/main/java/pro/taskana/user/internal/UserMapper.java [hunk 7]
```diff
@@ -27,12 +28,21 @@ public interface UserMapper {
   @Result(property = "data", column = "DATA")
   UserImpl findById(@Param("id") String id);
 
+  @SelectProvider(type = UserMapperSqlProvider.class, method = "findGroupsById")
+  Set<String> findGroupsById(@Param("id") String id);
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/user/internal/UserMapper.java:32]: the Param annotation necessary?

File: lib/taskana-core/src/main/java/pro/taskana/user/internal/UserMapperSqlProvider.java [hunk 8]
```diff
@@ -12,6 +12,9 @@ public class UserMapperSqlProvider {
   private static final String USER_INFO_VALUES =
       "#{id}, #{firstName}, #{lastName}, #{fullName}, #{longName}, #{email}, #{phone}, "
           + "#{mobilePhone}, #{orgLevel4}, #{orgLevel3}, #{orgLevel2}, #{orgLevel1}, #{data} ";
+
+  private static final String GROUP_INFO = "GROUP_INFO ";
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/user/internal/UserMapperSqlProvider.java:16]: Why did you move this to a constant?

File: lib/taskana-core/src/main/java/pro/taskana/workbasket/internal/WorkbasketQueryMapper.java [hunk 9]
```diff
@@ -31,7 +31,7 @@ public interface WorkbasketQueryMapper {
           + "MAX(PERM_CUSTOM_8::int) as MAX_CUSTOM_8, MAX(PERM_CUSTOM_9::int) as MAX_CUSTOM_9, MAX(PERM_CUSTOM_10::int) as MAX_CUSTOM_10, MAX(PERM_CUSTOM_11::int) as MAX_CUSTOM_11, MAX(PERM_CUSTOM_12::int) as MAX_CUSTOM_12 "
           + "</otherwise>"
           + "</choose>"
-          + "FROM WORKBASKET_ACCESS_LIST where ACCESS_ID IN (<if test='accessId != null'><foreach item='item' collection='accessId' separator=',' >#{item}</foreach></if>) group by WORKBASKET_ID ) a "
+          + "FROM WORKBASKET_ACCESS_LIST where LOWER(ACCESS_ID) IN (<if test='accessIds != null'><foreach item='item' collection='accessIds' separator=',' >#{item}</foreach></if>) group by WORKBASKET_ID ) a "
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/workbasket/internal/WorkbasketQueryMapper.java:34]: This is a "dirty fix" of the underlying issue: the input accessIds are not lowercase.

File: lib/taskana-test-api/src/main/java/pro/taskana/testapi/DefaultTestEntities.java [hunk 10]
```diff
@@ -47,6 +48,11 @@ public static ObjectReferenceBuilder defaultTestObjectReference() {
   public static UserBuilder randomTestUser() {
     return newUser()
         .id(UUID.randomUUID().toString().replace("-", ""))
+        .groups(
+            Set.of(
+                RandomStringGenerator.generateRandomString(30),
+                RandomStringGenerator.generateRandomString(30),
+                RandomStringGenerator.generateRandomString(30)))
```
Comment (mustaphazorgati) [inline lib/taskana-test-api/src/main/java/pro/taskana/testapi/DefaultTestEntities.java:55]: I don't think this is necessary :D

File: lib/taskana-test-api/src/main/java/pro/taskana/testapi/builder/UserBuilder.java [hunk 11]
```diff
@@ -22,6 +24,11 @@ public UserBuilder id(String id) {
     return this;
   }
 
+  public UserBuilder groups(Set<String> groups) {
```
Comment (mustaphazorgati) [inline lib/taskana-test-api/src/main/java/pro/taskana/testapi/builder/UserBuilder.java:27]: personally I'd use varargs here :)

File: rest/taskana-rest-spring/src/main/java/pro/taskana/common/rest/ldap/LdapClient.java [hunk 12]
```diff
@@ -633,7 +635,14 @@ class UserInfoContextMapper extends AbstractContextMapper<User> {
     @Override
     public User doMapFromContext(final DirContextOperations context) {
       final User user = new UserImpl();
+      String[] groupAttributes;
+      Set<String> groups =
+          (groupAttributes = context.getStringAttributes(getUserMemberOfGroupAttribute())) != null
+              ? Set.of(groupAttributes)
+              : Collections.emptySet();
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/common/rest/ldap/LdapClient.java:642]: Me no likey, due to nested assignment

File: rest/taskana-rest-spring/src/test/resources/application.properties [hunk 13]
```diff
@@ -1,16 +1,13 @@
 logging.level.pro.taskana=INFO
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/resources/application.properties:1]: revert changes

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

**Comment by mustaphazorgati** [comment]
open TODOs
* TaskanaEngineImpl#MINIMAL_TASKANA_SCHEMA_VERSION has to be set to 5.2.0
* UserRepresentationModel has to be expanded to understand groups
* DB script has to be tested for postgres

File: lib/taskana-core/src/main/java/pro/taskana/user/internal/UserMapperSqlProvider.java [hunk 14]
```diff
@@ -5,50 +5,62 @@
...
+        + "WHERE USER_ID = #{id} ";
   }
 
   public static String delete() {
-    return "DELETE FROM " + USER_INFO + WHERE_USER_ID;
+    return "DELETE FROM " + "USER_INFO " + "WHERE USER_ID = #{id} ";
+  }
+
+  public static String deleteGroups() {
+    return "DELETE FROM " + "GROUP_INFO " + "WHERE USER_ID = #{id} ";
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/user/internal/UserMapperSqlProvider.java:64]: join these strings :wink:

File: lib/taskana-core/src/main/java/pro/taskana/user/internal/UserServiceImpl.java [hunk 15]
```diff
@@ -52,7 +54,13 @@ public User getUser(String id) throws UserNotFoundException {
...
     }
 
-    user.setDomains(determineDomains(user.getId()));
+    Set<String> groups =
+        internalTaskanaEngine.executeInDatabaseConnection(() -> userMapper.findGroupsById(id));
+    user.setGroups(Objects.requireNonNullElse(groups, Collections.emptySet()));
+
+    Set<String> accessIds = new HashSet<>(groups);
+    accessIds.add(user.getId());
+    user.setDomains(determineDomains(accessIds.toArray(String[]::new)));
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/user/internal/UserServiceImpl.java:63]: pass the user object to `determineDomains`?

File: lib/taskana-core/src/main/java/pro/taskana/workbasket/internal/WorkbasketQueryImpl.java [hunk 16]
```diff
@@ -580,12 +579,10 @@ public void setUsedToAugmentTasks(boolean usedToAugmentTasks) {
   }
 
   public static void lowercaseAccessIds(String[] accessIdArray) {
-    if (TaskanaEngineConfiguration.shouldUseLowerCaseForAccessIds()) {
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/workbasket/internal/WorkbasketQueryImpl.java:583]: This if statement still needs to be part of the method.

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

File: lib/taskana-core/src/main/java/pro/taskana/TaskanaEngineConfiguration.java [hunk 17]
```diff
@@ -496,14 +504,12 @@ public void setSchemaName(String schemaName) {
...
-   * @return true if all access ids should be used in lower case, false otherwise
-   */
-  public static boolean shouldUseLowerCaseForAccessIds() {
-    return true;
+  public boolean isLowerCaseForAccessIds() {
+    return lowerCaseForAccessIds;
+  }
+
+  public void setLowerCaseForAccessIds(boolean lowerCaseForAccessIds) {
+    this.lowerCaseForAccessIds = lowerCaseForAccessIds;
```
[the remark this answers is not in the document]
Comment (norman-schmidt) [reply lib/taskana-core/src/main/java/pro/taskana/TaskanaEngineConfiguration.java:512]: They probably don't and you're right, maybe the complete impact of this change hasn't been verified enough. As it's not needed yet, i will revert the changes and only leave a test for the lower case accessIds.

File: lib/taskana-core/src/main/java/pro/taskana/user/internal/UserServiceImpl.java [hunk 18]
```diff
@@ -52,7 +54,13 @@ public User getUser(String id) throws UserNotFoundException {
       throw new UserNotFoundException(id);
     }
 
-    user.setDomains(determineDomains(user.getId()));
+    Set<String> groups =
+        internalTaskanaEngine.executeInDatabaseConnection(() -> userMapper.findGroupsById(id));
+    user.setGroups(Objects.requireNonNullElse(groups, Collections.emptySet()));
```
[the remark this answers is not in the document]
Comment (norman-schmidt) [reply lib/taskana-core/src/main/java/pro/taskana/user/internal/UserServiceImpl.java:59]: Wasn't sure if MyBatis maps the User it gets by 'findById' to a new User and also initializes it's fields. Could have just tested it :D

File: lib/taskana-core/src/main/java/pro/taskana/user/internal/UserServiceImpl.java (hunk 15 above)
Comment (mustaphazorgati) [reply lib/taskana-core/src/main/java/pro/taskana/user/internal/UserServiceImpl.java:63]: yep

Comment (mustaphazorgati) [reply lib/taskana-core/src/main/java/pro/taskana/user/internal/UserServiceImpl.java:63]: makes the call from outside a little cleaner and let's the method do its "dirty" work

Comment (norman-schmidt) [reply lib/taskana-core/src/main/java/pro/taskana/user/internal/UserServiceImpl.java:63]: alright!

**Review by mustaphazorgati — APPROVED** [review]
